internal/client/api/backend.go

```diff
@@ -352,3 +353,51 @@ type Backend[
 	// TODO: implement UsageInfo if we require it
 	// UsageInfo() *UsageInfo
 }
+
+// StorageProvider provides access to storage primitives
+type StorageProvider[H runtime.Hash, Hasher runtime.Hasher[H]] interface {
+	// Storage returns the value under the key in that block, given a blocks hash and a key.
+	Storage(hash H, key statemachine.StorageKey) (statemachine.StorageValue, error)
+
+	// StorageHash returns the value under the hash in that block, given a blocks hash and a key.
+	StorageHash(hash H, key statemachine.StorageKey) (*H, error)
+
+	// StorageKeys returns a [statemachine.KeysIter] that iterates over matching storage keys in that block
+	// given a blocks hash and a key prefix.
+	StorageKeys(hash H, prefix, startKey statemachine.StorageKey) (statemachine.KeysIter[H, Hasher], error)
```

Comment:
this should return an instance of the local `KeysIter`.  You need to create the constructor for `KeysIter`  which calls  `backend.RawIter` with a supplied `IterArgs` ([link to code](https://github.com/paritytech/polkadot-sdk/blob/030cb4a71b0b390626a586bfe7117b7c66b4700c/substrate/client/api/src/backend.rs#L330)).

Reply:
Based on [this comment](https://github.com/ChainSafe/gossamer/pull/4769#discussion_r2167582076) I've used `statemachine.KeysIter` and `statemachine.PairsIter` in `StorageProvider` and added the constructors to the `statemachine` package ([commit](https://github.com/ChainSafe/gossamer/pull/4769/commits/65615ce006a2865c7f63cf0adbb8fecf8af2d87c)).